File: src/encryption/c/aes/implementation.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <limits.h>

#include "implementation.h"
#include "../include/utils.h"
#include "../include/crypto_utils.h"
#include "aes_common.h"
#include "aes_key.h"
#include "aes_gcm.h"
#include "aes_cbc.h"
#include "aes_cfb.h"
#include "aes_ofb.h"

// register AES implementations - simplified to only register standard and custom
void register_aes_implementations(implementation_registry_t* registry) {
    int index = registry->count;
    int implementations_before = registry->count;
    
    // get the configuration to determine AES parameters
    char* key_size_str = getenv("AES_KEY_SIZE");
    char* mode_str = getenv("AES_MODE");
    char* use_stdlib_str = getenv("USE_STDLIB");
    char* use_custom_str = getenv("USE_CUSTOM");
    char* aes_enabled_str = getenv("AES_ENABLED");
    
    // default values if environment variables are not set
    int key_size = key_size_str ? atoi(key_size_str) : 256;  // default to 256
    char mode[16] = "GCM";  // default to GCM
    int use_stdlib = use_stdlib_str ? atoi(use_stdlib_str) : 1;  // default to true
    int use_custom = use_custom_str ? atoi(use_custom_str) : 0;  // default to false for now to match Go
    int aes_enabled = aes_enabled_str ? atoi(aes_enabled_str) : 1;  // Default to enabled
    
    // check if AES is enabled in the configuration
    if (!aes_enabled) {
        printf("AES implementations disabled in configuration\n");
        return;
    }
    
    if (mode_str) {
        strncpy(mode, mode_str, sizeof(mode) - 1);
    }
    
    // register standard AES implementation if enabled
    if (use_stdlib) {
        strcpy(registry->implementations[index].name, "aes");
        registry->implementations[index].algo_type = ALGO_AES;
        registry->implementations[index].is_custom = 0;
        registry->implementations[index].key_size = key_size;
        strcpy(registry->implementations[index].mode, mode);
        registry->implementations[index].init = aes_init;
        registry->implementations[index].cleanup = aes_cleanup;
        registry->implementations[index].generate_key = aes_generate_key;
        registry->implementations[index].encrypt = aes_encrypt;
        registry->implementations[index].decrypt = aes_decrypt;
        registry->implementations[index].encrypt_stream = aes_encrypt_stream;
        registry->implementations[index].decrypt_stream = aes_decrypt_stream;
        registry->count++;
    }
    
    // register custom AES implementation if enabled
    if (use_custom) {
        index = registry->count;
        strcpy(registry->implementations[index].name, "aes_custom");
        registry->implementations[index].algo_type = ALGO_AES;
        registry->implementations[index].is_custom = 1;
        registry->implementations[index].key_size = key_size;
        strcpy(registry->implementations[index].mode, mode);
        registry->implementations[index].init = aes_custom_init;
        registry->implementations[index].cleanup = aes_custom_cleanup;
        registry->implementations[index].generate_key = aes_custom_generate_key;
        registry->implementations[index].encrypt = aes_custom_encrypt;
        registry->implementations[index].decrypt = aes_custom_decrypt;
        registry->implementations[index].encrypt_stream = aes_encrypt_stream;
        registry->implementations[index].decrypt_stream = aes_decrypt_stream;
        registry->count++;
    }
    
    printf("Registered %d AES implementations\n", registry->count - implementations_before);
}

// standard library implementation functions
void* aes_init(void) {
    aes_context_t* context = (aes_context_t*)malloc(sizeof(aes_context_t));
    if (!context) {
        fprintf(stderr, "Error: Could not allocate AES context\n");
        return NULL;
    }
    
    memset(context, 0, sizeof(aes_context_t));
    
    // get configuration from environment variables
    char* key_size_str = getenv("AES_KEY_SIZE");
    char* mode_str = getenv("AES_MODE");
    
    // set key size from environment or default to 256
    context->key_size = key_size_str ? atoi(key_size_str) : 256;
    
    // set mode from environment or default to GCM
    if (mode_str) {
        strncpy(context->mode, mode_str, sizeof(context->mode) - 1);
    } else {
        strcpy(context->mode, "GCM");
    }
    
    context->is_custom = 0;
    context->key = NULL;
    context->key_length = 0;
    context->iv = NULL;
    context->iv_length = crypto_get_standard_iv_size("AES", context->mode);
    
    return context;
}

void aes_cleanup(void* context) {
    if (!context) return;
    
    aes_context_t* aes_context = (aes_context_t*)context;
    
    if (aes_context->key) {
        free(aes_context->key);
        aes_context->key = NULL;
    }
    
    if (aes_context->iv) {
        free(aes_context->iv);
        aes_context->iv = NULL;
    }
    
    free(aes_context);
}

unsigned char* aes_encrypt(void* context, const unsigned char* data, size_t data_length, const unsigned char* key, size_t* output_length) {
    if (!context || !data || data_length <= 0) return NULL;
    
    aes_context_t* aes_context = (aes_context_t*)context;
    
    // if key is provided, use it instead of the context key
    if (key) {
        if (aes_context->key) {
            free(aes_context->key);
        }
        
        aes_context->key_length = aes_context->key_size / 8;
        aes_context->key = (unsigned char*)malloc(aes_context->key_length);
        if (!aes_context->key) {
            fprintf(stderr, "Error: Could not allocate memory for AES key\n");
            return NULL;
        }
        
        memcpy(aes_context->key, key, aes_context->key_length);
    }
    
    // check if key exists
    if (!aes_context->key) {
        fprintf(stderr, "Error: AES key not set\n");
        return NULL;
    }
    
    // encrypt based on mode
    if (strcmp(aes_context->mode, "GCM") == 0) {
        #ifdef USE_OPENSSL
        int data_len = (int)data_length;
        int out_len = 0;
        unsigned char* result = aes_gcm_openssl_encrypt(aes_context, data, data_len, &out_len);
        if (output_length) *output_length = (size_t)out_len;
        return result;
        #else
        return aes_gcm_encrypt(aes_context, data, data_length, output_length);
        #endif
    } else if (strcmp(aes_context->mode, "CBC") == 0) {
        #ifdef USE_OPENSSL
        int data_len = (int)data_length;
        int out_len = 0;
        unsigned char* result = aes_cbc_openssl_encrypt(aes_context, data, data_len, &out_len);
        if (output_length) *output_length = (size_t)out_len;
        return result;
        #else
        int data_len = (int)data_length;
        int out_len = 0;
        unsigned char* result = aes_cbc_encrypt(aes_context, data, data_len, &out_len);
        if (output_length) *output_length = (size_t)out_len;
        return result;
        #endif
    } else if (strcmp(aes_context->mode, "CFB") == 0) {
        #ifdef USE_OPENSSL
        int data_len = (int)data_length;
        int out_len = 0;
        unsigned char* result = aes_cfb_openssl_encrypt(aes_context, data, data_len, &out_len);
        if (output_length) *output_length = (size_t)out_len;
        return result;
        #else
        int data_len = (int)data_length;
        int out_len = 0;
        unsigned char* result = aes_cfb_encrypt(aes_context, data, data_len, &out_len);
        if (output_length) *output_length = (size_t)out_len;
        return result;
        #endif
    } else if (strcmp(aes_context->mode, "OFB") == 0) {
        #ifdef USE_OPENSSL
        int data_len = (int)data_length;
        int out_len = 0;
        unsigned char* result = aes_ofb_openssl_encrypt(aes_context, data, data_len, &out_len);
        if (output_length) *output_length = (size_t)out_len;
        return result;
        #else
        int data_len = (int)data_length;
        int out_len = 0;
        unsigned char* result = aes_ofb_encrypt(aes_context, data, data_len, &out_len);
        if (output_length) *output_length = (size_t)out_len;
        return result;
        #endif
    } else {
        fprintf(stderr, "Error: Unsupported AES mode: %s\n", aes_context->mode);
        return NULL;
    }
}

unsigned char* aes_decrypt(void* context, const unsigned char* data, size_t data_length, const unsigned char* key, size_t* output_length) {
    if (!context || !data || data_length <= 0) return NULL;
    
    aes_context_t* aes_context = (aes_context_t*)context;
    
    // if key is provided, use it instead of the context key
    if (key) {
        if (aes_context->key) {
            free(aes_context->key);
        }
        
        aes_context->key_length = aes_context->key_size / 8;
        aes_context->key = (unsigned char*)malloc(aes_context->key_length);
        if (!aes_context->key) {
            fprintf(stderr, "Error: Could not allocate memory for AES key\n");
            return NULL;
        }
        
        memcpy(aes_context->key, key, aes_context->key_length);
    }
    
    // check if key exists
    if (!aes_context->key) {
        fprintf(stderr, "Error: AES key not set\n");
        return NULL;
    }
    
    // decrypt based on mode
    if (strcmp(aes_context->mode, "GCM") == 0) {
        #ifdef USE_OPENSSL
        int data_len = (int)data_length;
        int out_len = 0;
        unsigned char* result = aes_gcm_openssl_decrypt(aes_context, data, data_len, &out_len);
        if (output_length) *output_length = (size_t)out_len;
        return result;
        #else
        return aes_gcm_decrypt(aes_context, data, data_length, output_length);
        #endif
    } else if (strcmp(aes_context->mode, "CBC") == 0) {
        #ifdef USE_OPENSSL
        int data_len = (int)data_length;
        int out_len = 0;
        unsigned char* result = aes_cbc_openssl_decrypt(aes_context, data, data_len, &out_len);
        if (output_length) *output_length = (size_t)out_len;
        return result;
        #else
        return aes_cbc_decrypt(aes_context, data, data_length, output_length);
        #endif
    } else if (strcmp(aes_context->mode, "CFB") == 0) {
        #ifdef USE_OPENSSL
        int data_len = (int)data_length;
        int out_len = 0;
        unsigned char* result = aes_cfb_openssl_decrypt(aes_context, data, data_len, &out_len);
        if (output_length) *output_length = (size_t)out_len;
        return result;
        #else
        return aes_cfb_decrypt(aes_context, data, data_length, output_length);
        #endif
    } else if (strcmp(aes_context->mode, "OFB") == 0) {
        #ifdef USE_OPENSSL
        int data_len = (int)data_length;
        int out_len = 0;
        unsigned char* result = aes_ofb_openssl_decrypt(aes_context, data, data_len, &out_len);
        if (output_length) *output_length = (size_t)out_len;
        return result;
        #else
        return aes_ofb_decrypt(aes_context, data, data_length, output_length);
        #endif
    } else {
        fprintf(stderr, "Error: Unsupported AES mode: %s\n", aes_context->mode);
        return NULL;
    }
}



// Custom implementation functions
void* aes_custom_init(void) {
    aes_context_t* context = (aes_context_t*)malloc(sizeof(aes_context_t));
    if (!context) {
        fprintf(stderr, "Error: Could not allocate AES context\n");
        return NULL;
    }
    
    memset(context, 0, sizeof(aes_context_t));
    
    // get configuration from environment variables
    char* key_size_str = getenv("AES_KEY_SIZE");
    char* mode_str = getenv("AES_MODE");
    
    // set key size from environment or default to 256
    context->key_size = key_size_str ? atoi(key_size_str) : 256;
    
    // set mode from environment or default to GCM
    if (mode_str) {
        strncpy(context->mode, mode_str, sizeof(context->mode) - 1);
    } else {
        strcpy(context->mode, "GCM");
    }
    
    context->is_custom = 1;
    context->key = NULL;
    context->key_length = 0;
    context->iv = NULL;
    context->iv_length = crypto_get_standard_iv_size("AES", context->mode);
    
    return context;
}

void aes_custom_cleanup(void* context) {
    aes_cleanup(context); // reuse standard cleanup
}

unsigned char* aes_custom_encrypt(void* context, const unsigned char* data, size_t data_length, const unsigned char* key, size_t* output_length) {
    if (!context || !data || data_length <= 0) return NULL;
    
    aes_context_t* aes_context = (aes_context_t*)context;
    
    // if key is provided, use it instead of the context key
    if (key) {
        if (aes_context->key) {
            free(aes_context->key);
        }
        
        aes_context->key_length = aes_context->key_size / 8;
        aes_context->key = (unsigned char*)malloc(aes_context->key_length);
        if (!aes_context->key) {
            fprintf(stderr, "Error: Could not allocate memory for AES key\n");
            return NULL;
        }
        
        memcpy(aes_context->key, key, aes_context->key_length);
    }
    
    // check if key exists
    if (!aes_context->key) {
        fprintf(stderr, "Error: AES key not set\n");
        return NULL;
    }
    
    // encrypt based on mode using custom implementation
    if (strcmp(aes_context->mode, "GCM") == 0) {
        return aes_gcm_custom_encrypt(aes_context, data, data_length, output_length);
    } else if (strcmp(aes_context->mode, "CBC") == 0) {
        int data_len = (int)data_length;
        int out_len = 0;
        unsigned char* result = aes_cbc_custom_encrypt(aes_context, data, data_len, &out_len);
        if (output_length) *output_length = (size_t)out_len;
        return result;
    } else if (strcmp(aes_context->mode, "CFB") == 0) {
        int data_len = (int)data_length;
        int out_len = 0;
        unsigned char* result = aes_cfb_custom_encrypt(aes_context, data, data_len, &out_len);
        if (output_length) *output_length = (size_t)out_len;
        return result;
    } else if (strcmp(aes_context->mode, "OFB") == 0) {
        int data_len = (int)data_length;
        int out_len = 0;
        unsigned char* result = aes_ofb_custom_encrypt(aes_context, data, data_len, &out_len);
        if (output_length) *output_length = (size_t)out_len;
        return result;
    } else {
        fprintf(stderr, "Error: Unsupported AES mode: %s\n", aes_context->mode);
        return NULL;
    }
}

unsigned char* aes_custom_decrypt(void* context, const unsigned char* data, size_t data_length, const unsigned char* key, size_t* output_length) {
    if (!context || !data || data_length <= 0) return NULL;
    
    aes_context_t* aes_context = (aes_context_t*)context;
    
    // if key is provided, use it instead of the context key
    if (key) {
        if (aes_context->key) {
            free(aes_context->key);
        }
        
        aes_context->key_length = aes_context->key_size / 8;
        aes_context->key = (unsigned char*)malloc(aes_context->key_length);
        if (!aes_context->key) {
            fprintf(stderr, "Error: Could not allocate memory for AES key\n");
            return NULL;
        }
        
        memcpy(aes_context->key, key, aes_context->key_length);
    }
    
    // check if key exists
    if (!aes_context->key) {
        fprintf(stderr, "Error: AES key not set\n");
        return NULL;
    }
    
    // decrypt based on mode using custom implementation
    if (strcmp(aes_context->mode, "GCM") == 0) {
        return aes_gcm_custom_decrypt(aes_context, data, data_length, output_length);
    } else if (strcmp(aes_context->mode, "CBC") == 0) {
        return aes_cbc_custom_decrypt(aes_context, data, data_length, output_length);
    } else if (strcmp(aes_context->mode, "CFB") == 0) {
        return aes_cfb_custom_decrypt(aes_context, data, data_length, output_length);
    } else if (strcmp(aes_context->mode, "OFB") == 0) {
        return aes_ofb_custom_decrypt(aes_context, data, data_length, output_length);
    } else {
        fprintf(stderr, "Error: Unsupported AES mode: %s\n", aes_context->mode);
        return NULL;
    }
}

unsigned char* aes_encrypt_stream(void* context, const unsigned char* data, int data_length, const unsigned char* key, int chunk_index, int* output_length) {
    size_t data_len = (size_t)data_length;
    size_t out_len = 0;
    unsigned char* result = aes_encrypt(context, data, data_len, key, &out_len);
    if (output_length) *output_length = (int)out_len;
    return result;
}

unsigned char* aes_decrypt_stream(void* context, const unsigned char* data, int data_length, const unsigned char* key, int chunk_index, int* output_length) {
    size_t data_len = (size_t)data_length;
    size_t out_len = 0;
    unsigned char* result = aes_decrypt(context, data, data_len, key, &out_len);
    if (output_length) *output_length = (int)out_len;
    return result;
}

File: aes/implementation.zig
const std = @import("std");
const print = std.debug.print;
const Allocator = std.mem.Allocator;

// Import our modules
const utils = @import("../include/utils.zig");
const crypto_utils = @import("../include/crypto_utils.zig");

// Import AES mode implementations
const aes_common = @import("aes_common.zig");
const aes_key = @import("aes_key.zig");
const aes_gcm = @import("aes_gcm.zig");
const aes_cbc = @import("aes_cbc.zig");
const aes_ctr = @import("aes_ctr.zig");
const aes_ecb = @import("aes_ecb.zig");

// Import from parent module
const ImplementationRegistry = @import("../zig_core.zig").ImplementationRegistry;
const ImplementationInfo = @import("../zig_core.zig").ImplementationInfo;
const AlgorithmType = @import("../zig_core.zig").AlgorithmType;

// AES context structure
pub const AesContext = struct {
    key_size: i32,          // Key size in bits (128, 192, or 256)
    mode: [16]u8,           // Mode of operation (CBC, CTR, GCM, ECB)
    is_custom: bool,        // Whether this is a custom implementation
    key: ?[]u8,             // Encryption key
    key_length: usize,      // Key length in bytes
    iv: ?[]u8,              // Initialization vector
    iv_length: usize,       // IV length in bytes
    allocator: Allocator,   // Memory allocator
    
    pub fn init(allocator: Allocator) AesContext {
        return AesContext{
            .key_size = 256,
            .mode = std.mem.zeroes([16]u8),
            .is_custom = false,
            .key = null,
            .key_length = 0,
            .iv = null,
            .iv_length = 0,
            .allocator = allocator,
        };
    }
    
    pub fn deinit(self: *AesContext) void {
        if (self.key) |key| {
            crypto_utils.cryptoSecureFree(self.allocator, key);
            self.key = null;
        }
        if (self.iv) |iv| {
            self.allocator.free(iv);
            self.iv = null;
        }
    }
    
    pub fn setMode(self: *AesContext, mode: []const u8) void {
        @memset(&self.mode, 0);
        @memcpy(self.mode[0..mode.len], mode);
    }
    
    pub fn getMode(self: *const AesContext) []const u8 {
        return std.mem.sliceTo(&self.mode, 0);
    }
};

// Function to register AES implementations
pub fn registerAesImplementations(registry: *ImplementationRegistry, config_json: ?[]const u8, allocator: Allocator) !void {
    const implementations_before = registry.count;
    
    // Parse configuration from JSON if provided
    var key_size: i32 = 256; // Default to 256
    var mode_buf: [16]u8 = std.mem.zeroes([16]u8); // Buffer for mode string
    @memcpy(mode_buf[0.."GCM".len], "GCM"); // Default to GCM
    var use_stdlib = true; // Default to true
    var use_custom = true; // Default to true
    var aes_enabled = true; // Default to enabled
    
    if (config_json) |json_str| {
        if (std.json.parseFromSlice(std.json.Value, allocator, json_str, .{})) |parsed| {
            defer parsed.deinit();
            
            // Check if AES is enabled
            if (parsed.value.object.get("encryption_methods")) |encryption_methods| {
                if (encryption_methods.object.get("aes")) |aes_config| {
                    if (aes_config.object.get("enabled")) |enabled| {
                        aes_enabled = enabled.bool;
                    }
                    
                    if (aes_config.object.get("key_size")) |key_size_str| {
                        const key_size_value = std.fmt.parseInt(i32, key_size_str.string, 10) catch 256;
                        key_size = key_size_value;
                    }
                    
                    if (aes_config.object.get("mode")) |mode_str| {
                        @memset(&mode_buf, 0);
                        const copy_len = @min(mode_str.string.len, mode_buf.len - 1);
                        @memcpy(mode_buf[0..copy_len], mode_str.string[0..copy_len]);
                    }
                }
            }
            
            // Check test parameters for stdlib/custom settings
            if (parsed.value.object.get("test_parameters")) |test_params| {
                if (test_params.object.get("use_stdlib")) |stdlib| {
                    use_stdlib = stdlib.bool;
                }
                
                if (test_params.object.get("use_custom")) |custom| {
                    use_custom = custom.bool;
                }
            }
        } else |err| {
            print("Warning: Could not parse config JSON for AES: {}\n", .{err});
            // Continue with defaults
        }
    }
    
    // Check if AES is enabled in the configuration
    if (!aes_enabled) {
        return;
    }
    
    // Register standard AES implementation if enabled
    if (use_stdlib) {
        var impl = ImplementationInfo.init();
        @memcpy(impl.name[0.."aes".len], "aes");
        impl.algo_type = .aes;
        impl.is_custom = false;
        impl.key_size = key_size;
        const mode_slice = std.mem.sliceTo(&mode_buf, 0);
        const mode_copy_len = @min(mode_slice.len, impl.mode.len - 1);
        @memcpy(impl.mode[0..mode_copy_len], mode_slice[0..mode_copy_len]);
        impl.mode[mode_copy_len] = 0;
        impl.init_fn = aesInit;
        impl.cleanup_fn = aesCleanup;
        impl.generate_key_fn = aesGenerateKey;
        impl.encrypt_fn = aesEncrypt;
        impl.decrypt_fn = aesDecrypt;
        impl.encrypt_stream_fn = aesEncryptStream;
        impl.decrypt_stream_fn = aesDecryptStream;
        
        try registry.register(impl);
    }
    
    // Register custom AES implementation if enabled
    if (use_custom) {
        var impl = ImplementationInfo.init();
        @memcpy(impl.name[0.."aes_custom".len], "aes_custom");
        impl.algo_type = .aes;
        impl.is_custom = true;
        impl.key_size = key_size;
        const mode_slice = std.mem.sliceTo(&mode_buf, 0);
        const mode_copy_len = @min(mode_slice.len, impl.mode.len - 1);
        @memcpy(impl.mode[0..mode_copy_len], mode_slice[0..mode_copy_len]);
        impl.mode[mode_copy_len] = 0;
        impl.init_fn = aesCustomInit;
        impl.cleanup_fn = aesCustomCleanup;
        impl.generate_key_fn = aesCustomGenerateKey;
        impl.encrypt_fn = aesCustomEncrypt;
        impl.decrypt_fn = aesCustomDecrypt;
        impl.encrypt_stream_fn = aesEncryptStream;
        impl.decrypt_stream_fn = aesDecryptStream;
        
        try registry.register(impl);
    }
    
    print("Registered {d} AES implementations\n", .{registry.count - implementations_before});
}

// Standard library implementation functions
fn aesInit(allocator: Allocator) anyerror!*anyopaque {
    const context = try allocator.create(AesContext);
    context.* = AesContext.init(allocator);
    context.key_size = 256;
    context.setMode("GCM");
    context.is_custom = false;
    
    return @ptrCast(context);
}

fn aesCleanup(context: *anyopaque, allocator: Allocator) void {
    const aes_context: *AesContext = @ptrCast(@alignCast(context));
    aes_context.deinit();
    allocator.destroy(aes_context);
}

fn aesGenerateKey(context: *anyopaque, allocator: Allocator, key_length: *i32) anyerror![]u8 {
    const aes_context: *AesContext = @ptrCast(@alignCast(context));
    const key_size_bytes = @as(usize, @intCast(@divExact(aes_context.key_size, 8)));
    
    const key = try crypto_utils.cryptoGenerateKey(allocator, key_size_bytes);
    key_length.* = @intCast(key_size_bytes);
    
    return key;
}

fn aesEncrypt(context: *anyopaque, data: []const u8, key: []const u8, allocator: Allocator, output_length: *i32) anyerror![]u8 {
    const aes_context: *AesContext = @ptrCast(@alignCast(context));
    
    // Set the key if provided
    if (key.len > 0) {
        if (aes_context.key) |old_key| {
            crypto_utils.cryptoSecureFree(allocator, old_key);
        }
        
        aes_context.key_length = @as(usize, @intCast(@divExact(aes_context.key_size, 8)));
        aes_context.key = try allocator.dupe(u8, key[0..aes_context.key_length]);
    }
    
    // Check if key exists
    if (aes_context.key == null) {
        return error.AesKeyNotSet;
    }
    
    const mode = aes_context.getMode();
    
    // Encrypt based on mode
    if (std.mem.eql(u8, mode, "GCM")) {
        return aes_gcm.encrypt(aes_context, data, allocator, output_length);
    } else if (std.mem.eql(u8, mode, "CBC")) {
        return aes_cbc.encrypt(aes_context, data, allocator, output_length);
    } else if (std.mem.eql(u8, mode, "CTR")) {
        return aes_ctr.encrypt(aes_context, data, allocator, output_length);
    } else if (std.mem.eql(u8, mode, "ECB")) {
        return aes_ecb.encrypt(aes_context, data, allocator, output_length);
    } else {
        print("Error: Unsupported AES mode: {s}\n", .{mode});
        return error.UnsupportedAesMode;
    }
}

fn aesDecrypt(context: *anyopaque, data: []const u8, key: []const u8, allocator: Allocator, output_length: *i32) anyerror![]u8 {
    const aes_context: *AesContext = @ptrCast(@alignCast(context));
    
    // Set the key if provided
    if (key.len > 0) {
        if (aes_context.key) |old_key| {
            crypto_utils.cryptoSecureFree(allocator, old_key);
        }
        
        aes_context.key_length = @as(usize, @intCast(@divExact(aes_context.key_size, 8)));
        aes_context.key = try allocator.dupe(u8, key[0..aes_context.key_length]);
    }
    
    // Check if key exists
    if (aes_context.key == null) {
        return error.AesKeyNotSet;
    }
    
    const mode = aes_context.getMode();
    
    // Decrypt based on mode
    if (std.mem.eql(u8, mode, "GCM")) {
        return aes_gcm.decrypt(aes_context, data, allocator, output_length);
    } else if (std.mem.eql(u8, mode, "CBC")) {
        return aes_cbc.decrypt(aes_context, data, allocator, output_length);
    } else if (std.mem.eql(u8, mode, "CTR")) {
        return aes_ctr.decrypt(aes_context, data, allocator, output_length);
    } else if (std.mem.eql(u8, mode, "ECB")) {
        return aes_ecb.decrypt(aes_context, data, allocator, output_length);
    } else {
        print("Error: Unsupported AES mode: {s}\n", .{mode});
        return error.UnsupportedAesMode;
    }
}

// Custom implementation functions (simplified - same as standard for now)
fn aesCustomInit(allocator: Allocator) anyerror!*anyopaque {
    const context = try allocator.create(AesContext);
    context.* = AesContext.init(allocator);
    context.key_size = 256;
    context.setMode("GCM");
    context.is_custom = true;
    
    return @ptrCast(context);
}

fn aesCustomCleanup(context: *anyopaque, allocator: Allocator) void {
    const aes_context: *AesContext = @ptrCast(@alignCast(context));
    aes_context.deinit();
    allocator.destroy(aes_context);
}

fn aesCustomGenerateKey(context: *anyopaque, allocator: Allocator, key_length: *i32) anyerror![]u8 {
    return aesGenerateKey(context, allocator, key_length);
}

fn aesCustomEncrypt(context: *anyopaque, data: []const u8, key: []const u8, allocator: Allocator, output_length: *i32) anyerror![]u8 {
    // For now, custom implementation is the same as standard
    // In a real implementation, this would use a custom AES implementation
    return aesEncrypt(context, data, key, allocator, output_length);
}

fn aesCustomDecrypt(context: *anyopaque, data: []const u8, key: []const u8, allocator: Allocator, output_length: *i32) anyerror![]u8 {
    // For now, custom implementation is the same as standard
    // In a real implementation, this would use a custom AES implementation
    return aesDecrypt(context, data, key, allocator, output_length);
}

// Stream processing functions
fn aesEncryptStream(context: *anyopaque, data: []const u8, key: []const u8, chunk_index: i32, allocator: Allocator, output_length: *i32) anyerror![]u8 {
    _ = chunk_index; // For now, we'll treat stream processing the same as regular processing
    return aesEncrypt(context, data, key, allocator, output_length);
}

fn aesDecryptStream(context: *anyopaque, data: []const u8, key: []const u8, chunk_index: i32, allocator: Allocator, output_length: *i32) anyerror![]u8 {
    _ = chunk_index; // For now, we'll treat stream processing the same as regular processing
    return aesDecrypt(context, data, key, allocator, output_length);
} 